File: code/design-patterns/behavioral/chain-of-responsibility/go/chain_of_responsibility.go
package main

import (
	"fmt"
)

// Patient struct
type Patient struct {
	Insurance bool
	Broke     bool
	Log       string
}

// Handler interface
type Handler interface {
	Handle(patient *Patient)
	SetNext(handler Handler) Handler
}

// Department struct
type Department struct {
	next Handler
}

func (d *Department) SetNext(handler Handler) Handler {
	d.next = handler
	return handler
}

func (d *Department) Handle(patient *Patient) {
	if d.next != nil {
		d.next.Handle(patient)
	}
}

// Reception struct
type Reception struct {
	Department
}

func (r *Reception) Handle(patient *Patient) {
	if !patient.Insurance {
		fmt.Println("Ending chain - no insurance")
		return
	}
	patient.Log += "Reception -> "
	r.Department.Handle(patient)
}

// Doctor struct
type Doctor struct {
	Department
}

func (d *Doctor) Handle(patient *Patient) {
	patient.Log += "Doctor -> "
	d.Department.Handle(patient)
}

// MedicalExaminationRoom struct
type MedicalExaminationRoom struct {
	Department
}

func (m *MedicalExaminationRoom) Handle(patient *Patient) {
	patient.Log += "MedicalExaminationRoom -> "
	m.Department.Handle(patient)
}

// Cashier struct
type Cashier struct {
	Department
}

func (c *Cashier) Handle(patient *Patient) {
	if patient.Broke {
		fmt.Println("Ending chain - patient is broke")
		return
	}
	patient.Log += "Cashier -> "
	c.Department.Handle(patient)
}

File: code/design-patterns/behavioral/chain-of-responsibility/python/chain_of_responsibility.py
from abc import ABC, abstractmethod


class Handler(ABC):

    @abstractmethod
    def handle(self, patient):
        pass

    @abstractmethod
    def set_next(self, handler):
        pass


class Department(Handler):

    def __init__(self):
        self.next = None

    def set_next(self, handler):
        self.next = handler
        return handler

    def handle(self, patient):
        if self.next:
            return self.next.handle(patient)


class Reception(Department):

    def handle(self, patient):
        # Business logic
        if not patient.get("insurance"):
            print("Ending chain no insurance")
            return
        patient["log"] += "Reception -> "
        super().handle(patient)  # Call the next handler


class Doctor(Department):

    def handle(self, patient):
        patient["log"] += "Doctor -> "
        super().handle(patient)  # Call the next handler


class MedicalExaminationRoom(Department):

    def handle(self, patient):
        patient["log"] += "MedicalExaminationRoom -> "
        super().handle(patient)  # Call the next handler


class Cashier(Department):

    def handle(self, patient):
        if patient.get("broke"):
            print("Ending chain patient is broke")
            return
        patient["log"] += "Cashier -> "
        super().handle(patient)  # Call the next handler
